class_name TutorialEntitySpawner
extends EnemyPlacer

const ENEMY_CREDIT = 10

var _tutorial_stages: TutorialStages

@onready var _spawn_position: Marker2D = $SpawnPosition


func _ready():
	Events.enemy_died.connect(_on_enemy_died)


func setup(__player: Node2D):
	printerr("depricated: use setup_spawner")


func setup_spawner(player: Node2D, tutorial_stages: TutorialStages):
	_player = player
	_tutorial_stages = tutorial_stages

	_tutorial_stages.stage_started.connect(_on_tutorial_stage_started)
	_tutorial_stages.tutorial_event.connect(_on_tutorial_event)


func _on_tutorial_stage_started(stage: TutorialStages.Stages) -> void:
	match stage:
		TutorialStages.Stages.KILL_ENEMY:
			_spawn_enemy()


func _on_tutorial_event(event: TutorialStages.TutorialEvents) -> void:
	var has_enemy := get_child_count() > 2
	if (event == TutorialStages.TutorialEvents.SMELTER_LOADED) and (not has_enemy):
		_spawn_enemy()


func _on_enemy_died(enemy):
	_spawn_enemy_corps(enemy)

	_tutorial_stages.tutorial_event.emit(TutorialStages.TutorialEvents.ENEMY_DIED)


func _spawn_enemy():
	if _player == null:
		printerr("no player found")
		return

	var enemies = _monster_builder.create_monster_wave(ENEMY_CREDIT)

	for enemy in enemies:
		enemy.global_position = _spawn_position.position
		enemy.max_health = 30

		add_child(enemy)

		break  # we want only 1 enemy
